### PR TITLE
FiLM-Conditioned Fore-Foil SRF: Shape-Aware p_tan Correction

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The fore-foil (NACA6416 in OOD tandem cases) has a distinct DSDF1 gradient profile that differs from training shapes. Prior unconditioned fore-foil SRF attempts (#2117, #2124) failed because they applied a fixed correction bias regardless of which foil shape was present — causing overfitting to training shapes and regressing p_tan.

**Key insight from the aft-foil SRF win:** The aft-foil SRF head succeeded because it is FiLM-conditioned on gap/stagger geometry. Conditioning lets it learn geometry-specific corrections rather than a fixed bias. Apply the same principle to the fore-foil, conditioned on DSDF1 magnitude statistics (mean and std over fore-foil surface nodes) — a compact fingerprint distinguishing NACA6416 (OOD) from training shapes.

DSDF1 surface statistics encode shape identity: NACA6416 has a specific curvature profile visible in how DSDF1 magnitudes vary across its surface. By conditioning the fore-foil correction head on these statistics (plus gap/stagger), we give it the shape-aware signal the unconditioned versions lacked.

### Step 1: Add ForeFoilRefinementHead class

Near the `AftFoilRefinementHead` class in train.py, add:

```python
class ForeFoilRefinementHead(nn.Module):
    """FiLM-conditioned surface refinement for the fore-foil.
    Conditioning: 4D vector [dsdf1_mean, dsdf1_std, gap, stagger].
    """
    def __init__(self, n_hidden, n_out=4, hidden=192, n_layers=3, film_dim=4, film_hidden=32):
        super().__init__()
        self.film_mlp = nn.Sequential(
            nn.Linear(film_dim, film_hidden), nn.SiLU(),
            nn.Linear(film_hidden, hidden * 2),  # gamma and beta
        )
        layers = [nn.Linear(n_hidden + n_out, hidden), nn.GELU()]
        for _ in range(n_layers - 1):
            layers += [nn.Linear(hidden, hidden), nn.GELU()]
        self.main = nn.Sequential(*layers)
        self.out = nn.Linear(hidden, n_out)
        nn.init.zeros_(self.out.weight)
        nn.init.zeros_(self.out.bias)

    def forward(self, h, pred, film_cond):
        film = self.film_mlp(film_cond)
        gamma, beta = film.chunk(2, dim=-1)
        x = torch.cat([h, pred], dim=-1)
        x = self.main(x)
        x = x * (1 + gamma) + beta  # FiLM modulation
        return self.out(x)
```

### Step 2: Add config flags

```python
# Config dataclass:
fore_foil_srf: bool = False
fore_foil_srf_hidden: int = 192
fore_foil_srf_layers: int = 3

# argparse:
parser.add_argument('--fore_foil_srf', action='store_true')
parser.add_argument('--fore_foil_srf_hidden', type=int, default=192)
parser.add_argument('--fore_foil_srf_layers', type=int, default=3)
```

### Step 3: Instantiate head near aft_srf_head instantiation

```python
fore_srf_head = None
if cfg.fore_foil_srf:
    fore_srf_head = ForeFoilRefinementHead(
        n_hidden=_n_hidden, n_out=4,
        hidden=cfg.fore_foil_srf_hidden,
        n_layers=cfg.fore_foil_srf_layers,
        film_dim=4, film_hidden=32,
    ).to(device)
    base_opt.add_param_group({'params': list(fore_srf_head.parameters()), 'lr': _base_lr})
```

### Step 4: Compute fore-foil mask and FiLM conditioning in training loop

Add this alongside the `_aft_foil_mask` computation. You need access to `x` BEFORE standardization — store it as `x_raw = x.clone()` before `x = (x - stats["x_mean"]) / stats["x_std"]`.

```python
_fore_foil_mask = None
_fore_film_cond = None
if fore_srf_head is not None and model.training:
    _raw_saf_norm = x_raw[:, :, 2:6].norm(dim=-1)  # [B, N]
    _fore_foil_mask = is_surface & (_raw_saf_norm <= 0.005) & _is_tandem.unsqueeze(1)

    _dsdf1_mag = x_raw[:, :, 2:6].norm(dim=-1)
    _n_fore = _fore_foil_mask.float().sum(dim=1).clamp(min=1)
    _fore_dsdf1_mean = (_dsdf1_mag * _fore_foil_mask.float()).sum(dim=1) / _n_fore
    _fore_dsdf1_var  = ((_dsdf1_mag - _fore_dsdf1_mean.unsqueeze(1))**2 * _fore_foil_mask.float()).sum(dim=1) / _n_fore
    _fore_dsdf1_std  = _fore_dsdf1_var.sqrt()
    _fore_film_cond  = torch.stack([_fore_dsdf1_mean, _fore_dsdf1_std,
                                     _raw_gap_stagger[:, 0], _raw_gap_stagger[:, 1]], dim=-1)  # [B, 4]
```

### Step 5: Apply fore-foil correction after aft-foil correction block

```python
if fore_srf_head is not None and _fore_foil_mask is not None and _fore_foil_mask.any():
    fore_idx = _fore_foil_mask.nonzero(as_tuple=False)
    if fore_idx.numel() > 0:
        fore_hidden = hidden[fore_idx[:, 0], fore_idx[:, 1]]
        fore_pred   = pred[fore_idx[:, 0], fore_idx[:, 1]]
        _fore_cond  = _fore_film_cond[fore_idx[:, 0]]
        with torch.amp.autocast("cuda", dtype=torch.bfloat16):
            fore_correction = fore_srf_head(fore_hidden, fore_pred, _fore_cond).float()
        pred = pred.clone()
        pred[fore_idx[:, 0], fore_idx[:, 1]] += fore_correction
```

### Step 6: Run 2 seeds

```bash
# Seed 42
cd cfd_tandemfoil && python train.py --agent fern \
  --wandb_name "fern/fore-srf-film-s42" --wandb_group phase6/fore-foil-srf-film \
  --fore_foil_srf --fore_foil_srf_hidden 192 --fore_foil_srf_layers 3 --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias

# Seed 73: same flags, --seed 73, wandb_name "fern/fore-srf-film-s73"
```

Also log `_fore_foil_mask.sum()` at epoch 1 to confirm fore-foil nodes are being found.

## Baseline

| Metric | Target |
|--------|--------|
| p_in   | < 13.05 |
| p_oodc | < 7.70  |
| **p_tan** | **< 28.60** |
| p_re   | < 6.55  |

W&B baseline: d7l91p0x (s42), j9btfx09 (s73)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "<name>/baseline" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT fern:

## Results

### Summary

| Config | Seed | p_in | p_oodc | p_tan | p_re | VRAM | W&B |
|--------|------|------|--------|-------|------|------|-----|
| fore_srf (post-MLP FiLM) | 42 | 13.1 | 7.4 | 30.8 | 6.3 | 45.8GB | snmfs7bm |
| fore_srf (post-MLP FiLM) | 73 | 13.1 | 7.4 | 29.7 | 6.5 | 45.1GB | o692pxsi |
| **2-seed avg** | — | **13.10** | **7.40** | **30.25** | **6.40** | — | — |
| **Baseline** | — | 13.05 | 7.70 | 28.60 | 6.55 | — | d7l91p0x, j9btfx09 |

**Negative result on p_tan.** p_tan regressed significantly: +5.8% (30.25 vs 28.60). p_oodc improved slightly (-3.9%: 7.4 vs 7.70), but p_tan — the primary target — degraded.

### Commands used

```bash
# Seed 42
cd cfd_tandemfoil && python train.py --agent fern   --wandb_name "fern/fore-srf-film-s42" --wandb_group phase6/fore-foil-srf-film   --fore_foil_srf --fore_foil_srf_hidden 192 --fore_foil_srf_layers 3 --seed 42   --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4   --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96   --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5   --cosine_T_max 160 --pressure_first --pressure_deep   --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3   --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05   --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias

# Seed 73: same flags, --seed 73, --wandb_name "fern/fore-srf-film-s73"
```

### Training details

- Epochs: 149 (s42), 148 (s73) — wall-clock limit at 180 min
- Head params: ForeFoilRefinementHead = 125,155 params
- VRAM: ~45.1–45.8 GB

### Diagnostic: correction norm still collapses

The  metric shows the same pattern as the pre-MLP variant: initial activity (small █ at start) then collapse to near-zero (▁ throughout most of training). Applying FiLM post-MLP vs. pre-MLP does not resolve the fundamental issue.

### What happened

This is the **fourth consecutive failure** of fore-foil surface corrections across different architectures:
- #2117: Split SRF head (unconditioned) → p_tan +9-11% regression
- #2124: Stacked additive SRF head (unconditioned) → p_tan regression
- #2159: FiLM pre-MLP on hidden features → p_tan +2.4% regression, correction norm collapses
- **This PR**: FiLM post-MLP on MLP output → p_tan +5.8% regression, correction norm collapses

The correction norm pattern (high early, decays to zero) is consistent across all four attempts. The optimizer finds a better local minimum with zero correction than with any active correction — suggesting the main model + existing SRF head already handles fore-foil pressure as well as it can given the training distribution.

Note: s42 shows p_tan=30.8 (much worse than s73's 29.7), indicating high seed variance on this metric, which may also be obscuring any signal.

### Suggested follow-ups

1. **Stop fore-foil SRF experiments.** Four attempts with progressively more sophisticated conditioning have all failed. The correction consistently learns to be zero. This is strong evidence that the approach direction is exhausted.

2. **The p_oodc improvement** (7.4 vs 7.70) is interesting — fore-foil SRF may actually help OOD-conditions evaluation. If the advisor believes this direction has merit, a focused p_oodc experiment might be worthwhile.

3. **Focus on the main model.** The next gains are more likely to come from changes to the Transolver backbone itself (e.g., better cross-foil attention, more slices for tandem configurations) rather than additive correction heads.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2161 because ADVISOR: Closing — p_tan=30.25 (+5.8% vs baseline). This is the 4th consecutive failure of fore-foil SRF:

- #2117: Split SRF (unconditioned) → p_tan +9-11%
- #2124: Stacked additive SRF (unconditioned) → p_tan regression
- #2159: FiLM pre-MLP → p_tan +2.4%, correction collapses
- #2161: FiLM post-MLP → p_tan +5.8%, correction collapses

**Definitive conclusion: fore-foil surface refinement is a dead end.** The correction norm consistently collapses to zero across all 4 architectures — the optimizer finds a better minimum without any correction. The main model + existing aft_foil_srf already handles fore-foil pressure as well as it can given current input representation.

I agree with your analysis: the next gains come from changes to the representation or training dynamics, not additive correction heads.

Interesting: p_oodc improved (-3.9%), p_re improved (-2.3%). The fore-foil SRF acts as a regularizer for non-tandem-transfer metrics, but actively hurts the one metric we need most.

Great diagnostic work tracking the correction norm across all 4 attempts. Moving you to a new direction..
